spec/datadog/core/remote/dispatcher_spec.rb

```diff
@@ -0,0 +1,170 @@
+            }
+          ],
+          id: '874459ae-137f-4c99-9c54-109b1a117b86'
+        },
+        {
+          conditions: [
+            {
+              operator: 'match_regex',
+              parameters: {
+                inputs: [
+                  {
+                    address: 'server.request.uri.raw'
+                  }
+                ],
+                options: {
+                  case_sensitive: false
+                },
+                regex: '^/waf'
+              }
+            }
+          ],
+          id: 'd1390949-cf1a-408d-bc3f-043d0689d89e'
+        },
+        {
+          id: '5fe8e530-d3ec-4e6d-bc06-0a6637c6e763',
+          rules_target: [
+            {
+              rule_id: 'ua0-600-55x'
+            }
+          ]
+        },
+        {
+          conditions: [
+            {
+              operator: 'ip_match',
+              parameters: {
+                inputs: [
+                  {
+                    address: 'http.client_ip'
+                  }
+                ],
+                list: [
+                  '8.8.8.8'
+                ]
+              }
+            }
+          ],
+          id: '081e1fbe-c73b-4ad2-bb83-4752354271bc'
+        }
+      ],
+      rules_override: []
+    }
+  end
+  let(:string_io_content) { StringIO.new(raw.to_json) }
+  let(:content) do
+    Datadog::Core::Remote::Configuration::Content.parse({ :path => path.to_s, :content => string_io_content })
+  end
+  let(:repository) { Datadog::Core::Remote::Configuration::Repository.new }
+
+  let(:receiver_block) { proc { |_repository, _changes| } }
+
+  let(:receiver) do
+    described_class::Receiver.new(matcher, &receiver_block)
+  end
+
+  subject(:dispatcher) do
+    d = described_class.new
+    d.receivers << receiver
+    d
+  end
+
+  describe '#dispatch' do
+    context 'receiver matches' do
+      it 'dispatches #call on matched changes' do
+        changes = repository.transaction do |_repository, transaction|
+          transaction.insert(path, target, content)
+        end
+
+        expect(receiver).to receive(:call).with(
+          repository,
+          [instance_of(Datadog::Core::Remote::Configuration::Repository::Change::Inserted)]
+        )
+
+        dispatcher.dispatch(changes, repository)
+      end
+    end
+
+    context 'receiver does not matches' do
+      let(:matcher) do
+        described_class::Matcher.new do |_path|
+          false
+        end
+      end
+
+      it 'does not dispatches #call on non matched changes' do
+        changes = repository.transaction do |_repository, transaction|
+          transaction.insert(path, target, content)
+        end
+
+        expect(receiver).to_not receive(:call).with(
+          repository,
+          [instance_of(Datadog::Core::Remote::Configuration::Repository::Change::Inserted)]
+        )
+
+        dispatcher.dispatch(changes, repository)
+      end
+    end
+  end
+
+  describe Datadog::Core::Remote::Dispatcher::Matcher::Product do
```

Comment:
Minor note but... I::Think::We're::Getting::A::Bit::Too::Deep::With::All::The::Namespacing. Consider flattening things a bit maybe?